File: macos/AgentCord/CodexUsage.swift
//
//  CodexUsage.swift
//  AgentCord
//
//  Reads ChatGPT / Codex subscription rate limits through Codex's stable
//  `codex app-server` JSONL interface. Codex remains the sole owner of its
//  credentials and refresh-token lifecycle; AgentCord never reads or refreshes
//  OAuth tokens directly.
//

import Foundation
import Combine

final class CodexUsage: ObservableObject {

    /// Latest usage snapshot, or nil when it could not be fetched.
    @Published private(set) var current: CodexUsageInfo?

    /// True when app-server reports a ChatGPT-backed Codex account.
    @Published private(set) var isAuthenticated = false

    var pollInterval: TimeInterval = 300
    var minFetchInterval: TimeInterval = 60
    var maxStaleness: TimeInterval = 1800
    var requestTimeout: TimeInterval = 15

    private var lastSuccess: Date = .distantPast
    private var lastAttempt: Date = .distantPast

    private let queue = DispatchQueue(label: "com.agentcord.codex-usage", qos: .utility)
    private var timer: DispatchSourceTimer?

    /// Only one short-lived app-server probe may be active at a time.
    private var activeProcess: Process?
    private var activeInput: FileHandle?
    private var activeOutput: FileHandle?
    private var responseBuffer = Data()
    private var activeAuthenticated: Bool?
    private var requestGeneration = 0

    private static let cacheURL: URL = {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        let dir = base.appendingPathComponent("AgentCord", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("codex-usage-cache.json")
    }()

    init() {
        if let cached = Self.loadCache(), Date().timeIntervalSince(cached.fetchedAt) <= maxStaleness {
            current = cached.info
            lastSuccess = cached.fetchedAt
        }
    }

    func start() {
        let t = DispatchSource.makeTimerSource(queue: queue)
        let firstDelay: TimeInterval = (current != nil) ? 5 : 2
        t.schedule(deadline: .now() + firstDelay, repeating: pollInterval)
        t.setEventHandler { [weak self] in self?.fetch() }
        t.resume()
        timer = t
    }

    func stop() {
        timer?.cancel()
        timer = nil
        queue.async { [weak self] in self?.cancelActiveRequest() }
    }

    func refresh() {
        queue.async { [weak self] in
            guard let self else { return }
            guard Date().timeIntervalSince(self.lastAttempt) >= self.minFetchInterval else { return }
            self.fetch()
        }
    }

    // MARK: - App-server fetch

    private func fetch() {
        guard activeProcess == nil else { return }
        let now = Date()
        guard now.timeIntervalSince(lastAttempt) >= minFetchInterval else { return }
        lastAttempt = now

        guard let executable = Self.codexExecutableURL() else {
            publishAuth(false)
            handleFailure()
            return
        }

        let process = Process()
        let input = Pipe()
        let output = Pipe()

        process.executableURL = executable
        process.arguments = ["app-server"]
        process.standardInput = input
        process.standardOutput = output
        // Never merge diagnostics into stdout: stdout is JSONL protocol data.
        process.standardError = FileHandle.nullDevice

        var environment = ProcessInfo.processInfo.environment
        if environment["CODEX_HOME"]?.isEmpty != false {
            environment["CODEX_HOME"] = FileManager.default.homeDirectoryForCurrentUser
                .appendingPathComponent(".codex", isDirectory: true).path
        }
        process.environment = environment

        requestGeneration += 1
        let generation = requestGeneration
        activeProcess = process
        activeInput = input.fileHandleForWriting
        activeOutput = output.fileHandleForReading
        activeAuthenticated = nil
        responseBuffer.removeAll(keepingCapacity: true)

        output.fileHandleForReading.readabilityHandler = { [weak self, weak process] handle in
            let data = handle.availableData
            guard !data.isEmpty, let self, let process else { return }
            self.queue.async {
                guard self.activeProcess === process, self.requestGeneration == generation else { return }
                self.responseBuffer.append(data)
                self.consumeResponseLines(from: process)
            }
        }

        process.terminationHandler = { [weak self, weak process] _ in
            guard let self, let process else { return }
            self.queue.async {
                guard self.activeProcess === process, self.requestGeneration == generation else { return }
                self.finishRequest(process, info: nil, authenticated: self.activeAuthenticated)
            }
        }

        do {
            try process.run()
        } catch {
            finishRequest(process, info: nil, authenticated: false)
            return
        }

        guard let messages = Self.appServerMessages() else {
            finishRequest(process, info: nil, authenticated: nil)
            return
        }
        input.fileHandleForWriting.write(messages)

        queue.asyncAfter(deadline: .now() + requestTimeout) { [weak self, weak process] in
            guard let self, let process,
                  self.activeProcess === process,
                  self.requestGeneration == generation else { return }
            self.finishRequest(process, info: nil, authenticated: self.activeAuthenticated)
        }
    }

    /// Parses complete JSONL records while retaining a partial final record.
    private func consumeResponseLines(from process: Process) {
        let newline = Data([0x0A])
        while let range = responseBuffer.range(of: newline) {
            let line = responseBuffer.subdata(in: responseBuffer.startIndex..<range.lowerBound)
            responseBuffer.removeSubrange(responseBuffer.startIndex...range.lowerBound)
            guard !line.isEmpty,
                  let object = try? JSONSerialization.jsonObject(with: line) as? [String: Any],
                  let id = object["id"] as? Int else { continue }

            switch id {
            case 1:
                if let envelope = try? JSONDecoder().decode(AccountEnvelope.self, from: line) {
                    let type = envelope.result?.account?.type
                    activeAuthenticated = type == "chatgpt" || type == "personalAccessToken"
                    publishAuth(activeAuthenticated == true)
                }
            case 2:
                guard object["error"] == nil,
                      let envelope = try? JSONDecoder().decode(RateLimitsEnvelope.self, from: line),
                      let result = envelope.result,
                      let info = result.toUsageInfo() else {
                    finishRequest(process, info: nil, authenticated: activeAuthenticated)
                    return
                }
                finishRequest(process, info: info, authenticated: true)
                return
            default:
                continue
            }
        }
    }

    private func finishRequest(
        _ process: Process, info: CodexUsageInfo?, authenticated: Bool?
    ) {
        guard activeProcess === process else { return }

        activeOutput?.readabilityHandler = nil
        activeInput?.closeFile()
        activeInput = nil
        activeOutput = nil
        activeProcess = nil
        activeAuthenticated = nil
        responseBuffer.removeAll(keepingCapacity: true)
        if process.isRunning { process.terminate() }

        if let authenticated { publishAuth(authenticated) }
        if let info {
            lastSuccess = Date()
            Self.saveCache(info, fetchedAt: lastSuccess)
            publish(info)
        } else {
            handleFailure()
        }
    }

    private func cancelActiveRequest() {
        guard let process = activeProcess else { return }
        activeOutput?.readabilityHandler = nil
        activeInput?.closeFile()
        activeInput = nil
        activeOutput = nil
        activeProcess = nil
        activeAuthenticated = nil
        responseBuffer.removeAll(keepingCapacity: true)
        if process.isRunning { process.terminate() }
    }

    private static func appServerMessages() -> Data? {
        let messages: [[String: Any]] = [
            [
                "method": "initialize",
                "id": 0,
                "params": [
                    "clientInfo": [
                        "name": "agentcord",
                        "title": "AgentCord",
                        "version": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0"
                    ]
                ]
            ],
            ["method": "initialized", "params": [:]],
            ["method": "account/read", "id": 1, "params": ["refreshToken": false]],
            ["method": "account/rateLimits/read", "id": 2]
        ]

        var payload = Data()
        for message in messages {
            guard let data = try? JSONSerialization.data(withJSONObject: message) else { return nil }
            payload.append(data)
            payload.append(0x0A)
        }
        return payload
    }

    /// GUI apps often receive a minimal PATH, so check the inherited PATH and
    /// the standard Codex install locations explicitly.
    private static func codexExecutableURL() -> URL? {
        let fm = FileManager.default
        var paths: [String] = []
        if let configured = ProcessInfo.processInfo.environment["CODEX_BINARY"], !configured.isEmpty {
            paths.append(configured)
        }
        if let path = ProcessInfo.processInfo.environment["PATH"] {
            paths.append(contentsOf: path.split(separator: ":").map { "\($0)/codex" })
        }

        let home = fm.homeDirectoryForCurrentUser.path
        paths.append(contentsOf: [
            "/opt/homebrew/bin/codex",
            "/usr/local/bin/codex",
            "\(home)/.local/bin/codex",
            "\(home)/bin/codex"
        ])

        for path in paths where fm.isExecutableFile(atPath: path) {
            return URL(fileURLWithPath: path)
        }
        return nil
    }

    // MARK: - State and cache

    private func handleFailure() {
        if Date().timeIntervalSince(lastSuccess) > maxStaleness {
            Self.clearCache()
            publish(nil)
        }
    }

    private func publish(_ info: CodexUsageInfo?) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if self.current != info { self.current = info }
        }
    }

    private func publishAuth(_ authenticated: Bool) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if self.isAuthenticated != authenticated { self.isAuthenticated = authenticated }
        }
    }

    private struct CachePayload: Codable {
        var fetchedAt: Date
        var info: CodexUsageInfo
    }

    private static func loadCache() -> CachePayload? {
        guard let data = try? Data(contentsOf: cacheURL) else { return nil }
        return try? JSONDecoder().decode(CachePayload.self, from: data)
    }

    private static func saveCache(_ info: CodexUsageInfo, fetchedAt: Date) {
        let payload = CachePayload(fetchedAt: fetchedAt, info: info)
        guard let data = try? JSONEncoder().encode(payload) else { return }
        try? data.write(to: cacheURL, options: .atomic)
    }

    private static func clearCache() {
        try? FileManager.default.removeItem(at: cacheURL)
    }
}

// MARK: - App-server wire format

private struct AccountEnvelope: Decodable {
    struct Result: Decodable {
        struct Account: Decodable { let type: String }
        let account: Account?
    }
    let result: Result?
}

private struct RateLimitsEnvelope: Decodable {
    struct Result: Decodable {
        struct Snapshot: Decodable {
            struct Window: Decodable {
                let usedPercent: Double?
                let windowDurationMins: Int?
                let resetsAt: Double?
            }

            let limitId: String?
            let limitName: String?
            let primary: Window?
            let secondary: Window?
            let planType: String?
            let rateLimitReachedType: String?
        }

        let rateLimits: Snapshot
        let rateLimitsByLimitId: [String: Snapshot]?

        func toUsageInfo() -> CodexUsageInfo? {
            guard let primary = rateLimits.primary else { return nil }
            let reached = rateLimits.rateLimitReachedType != nil
            let primaryLabel = Self.label(forMinutes: primary.windowDurationMins, fallback: "Primary limit")

            let secondary = rateLimits.secondary.map { Self.makeWindow($0, limitReached: false) }
            let secondaryLabel = rateLimits.secondary.map {
                Self.label(forMinutes: $0.windowDurationMins, fallback: "Secondary limit")
            }

            var additionalWindows: [CodexUsageInfo.NamedWindow] = []
            for (key, snapshot) in (rateLimitsByLimitId ?? [:]).sorted(by: { $0.key < $1.key }) {
                let id = snapshot.limitId ?? key
                guard id != (rateLimits.limitId ?? "codex"), id != "codex" else { continue }
                let baseLabel = Self.displayName(snapshot.limitName ?? id)
                if let window = snapshot.primary {
                    additionalWindows.append(.init(
                        id: "\(id)-primary",
                        label: baseLabel,
                        window: Self.makeWindow(window, limitReached: snapshot.rateLimitReachedType != nil)
                    ))
                }
                if let window = snapshot.secondary {
                    additionalWindows.append(.init(
                        id: "\(id)-secondary",
                        label: "\(baseLabel) · \(Self.label(forMinutes: window.windowDurationMins, fallback: "Secondary"))",
                        window: Self.makeWindow(window, limitReached: false)
                    ))
                }
            }

            return CodexUsageInfo(
                primary: Self.makeWindow(primary, limitReached: reached),
                primaryLabel: primaryLabel,
                secondary: secondary,
                secondaryLabel: secondaryLabel,
                planType: rateLimits.planType,
                additionalWindows: additionalWindows
            )
        }

        private static func makeWindow(_ window: Snapshot.Window, limitReached: Bool) -> UsageInfo.Window {
            let percent = min(100, max(0, Int((window.usedPercent ?? 0).rounded())))
            let severity: String
            if limitReached || percent >= 90 { severity = "critical" }
            else if percent >= 70 { severity = "warning" }
            else { severity = "normal" }
            let resetsAt = window.resetsAt.map { Date(timeIntervalSince1970: $0) }
            return UsageInfo.Window(percent: percent, severity: severity, resetsAt: resetsAt)
        }

        private static func label(forMinutes minutes: Int?, fallback: String) -> String {
            guard let minutes, minutes > 0 else { return fallback }
            if minutes <= 6 * 60 { return "5-hour session" }
            if minutes <= 8 * 24 * 60 { return "Weekly limit" }
            if minutes <= 40 * 24 * 60 { return "Monthly limit" }
            return fallback
        }

        private static func displayName(_ raw: String) -> String {
            raw.replacingOccurrences(of: "_", with: " ")
                .replacingOccurrences(of: "-", with: " ")
                .split(separator: " ")
                .map { $0.prefix(1).uppercased() + $0.dropFirst() }
                .joined(separator: " ")
        }
    }

    let result: Result?
}

File: windows/CodexUsage.cs
// Reads Codex / ChatGPT subscription rate limits through Codex's local
// app-server JSONL protocol. Codex remains the owner of its credentials and
// refresh lifecycle; AgentCord never reads OAuth tokens directly.
//
// Best-effort: a failed probe keeps the last disk-cached snapshot so relaunch
// / idle stretches still show numbers instead of "Waiting for Codex usage…".
// The cache is bound to Codex's local account_id so a logout or account switch
// cannot keep showing another account's numbers. We only read account_id for
// that binding — OAuth tokens stay owned by Codex / app-server.

using System.Diagnostics;
using System.IO;
using System.Text.Json;

namespace AgentCord;

public sealed class CodexUsage : IDisposable
{
    public CodexUsageInfo? Current { get; private set; }

    /// <summary>Email of the signed-in ChatGPT account, from account/read.</summary>
    public string? AccountEmail { get; private set; }

    public bool IsAuthenticated { get; private set; }

    public TimeSpan PollInterval { get; init; } = TimeSpan.FromMinutes(5);
    public TimeSpan MinFetchInterval { get; init; } = TimeSpan.FromMinutes(1);
    /// <summary>How long a disk-cached snapshot may still be shown after the last
    /// successful fetch. Matches macOS — generous so an idle stretch or a
    /// relaunch shows the last known numbers instead of "Waiting…".</summary>
    public TimeSpan MaxStaleness { get; init; } = TimeSpan.FromHours(24);
    public TimeSpan RequestTimeout { get; init; } = TimeSpan.FromSeconds(15);

    private readonly object _lock = new();
    private DateTime _lastAttempt = DateTime.MinValue;
    private DateTime _lastSuccess = DateTime.MinValue;
    private string? _accountKey;
    private System.Threading.Timer? _timer;
    private int _fetching;
    private bool _disposed;

    public CodexUsage()
    {
        var key = CurrentAccountKey();
        if (key is null)
        {
            ClearCache();
            return;
        }

        var cached = LoadCache();
        if (cached is not null
            && cached.AccountKey == key
            && DateTime.UtcNow - cached.FetchedAt <= MaxStaleness)
        {
            Current = cached.Info;
            _lastSuccess = cached.FetchedAt;
            _accountKey = key;
            IsAuthenticated = true;
        }
        else if (cached is not null)
        {
            ClearCache();
        }
    }

    public void Start() => SetEnabled(true);

    public void SetEnabled(bool enabled)
    {
        if (enabled)
        {
            if (_timer is not null) return;
            var first = Current is null ? TimeSpan.FromSeconds(2) : TimeSpan.FromSeconds(5);
            _timer = new System.Threading.Timer(_ => _ = FetchAsync(), null, first, PollInterval);
            return;
        }
        _timer?.Dispose();
        _timer = null;
    }

    public void Refresh()
    {
        lock (_lock)
        {
            if (DateTime.UtcNow - _lastAttempt < MinFetchInterval) return;
        }
        _ = FetchAsync();
    }

    public void Dispose()
    {
        _disposed = true;
        _timer?.Dispose();
    }

    private async Task FetchAsync()
    {
        if (_disposed || Interlocked.Exchange(ref _fetching, 1) == 1) return;
        lock (_lock) _lastAttempt = DateTime.UtcNow;

        try
        {
            var key = CurrentAccountKey();
            if (key is null)
            {
                InvalidateIdentity();
                return;
            }

            if (_accountKey is not null && _accountKey != key)
                InvalidateIdentity();
            _accountKey = key;

            var executable = FindExecutable();
            if (executable is null)
            {
                HandleFailure();
                return;
            }

            using var process = new Process
            {
                StartInfo = new ProcessStartInfo(executable)
                {
                    RedirectStandardInput = true,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                },
            };
            process.StartInfo.ArgumentList.Add("app-server");
            // Match macOS: point app-server at the same home CodexSession uses
            // so a custom CODEX_HOME is honored for credentials too.
            process.StartInfo.Environment["CODEX_HOME"] = CodexPaths.ResolveHome();
            if (!process.Start())
            {
                HandleFailure();
                return;
            }

            await process.StandardInput.WriteLineAsync(
                """{"method":"initialize","id":0,"params":{"clientInfo":{"name":"agentcord","title":"AgentCord","version":"0.4.0"}}}""");
            await process.StandardInput.WriteLineAsync(
                """{"method":"account/read","id":1,"params":{"refreshToken":false}}""");
            await process.StandardInput.WriteLineAsync(
                """{"method":"account/rateLimits/read","id":2,"params":null}""");
            await process.StandardInput.FlushAsync();

            using var timeout = new CancellationTokenSource(RequestTimeout);
            CodexUsageInfo? result = null;
            string? email = null;
            bool? authenticated = null;
            try
            {
                while (!timeout.IsCancellationRequested)
                {
                    var line = await process.StandardOutput.ReadLineAsync(timeout.Token);
                    if (line is null) break;
                    if (TryParseAccount(line, out var accountEmail, out var accountAuth))
                    {
                        email = accountEmail;
                        authenticated = accountAuth;
                        continue;
                    }
                    if (TryParseResponse(line, out result)) break;
                }
            }
            catch (OperationCanceledException)
            {
                // The app-server is intentionally short-lived from our point
                // of view; a timeout simply retains the last good snapshot.
            }

            if (!process.HasExited)
            {
                try { process.Kill(entireProcessTree: true); }
                catch { }
            }

            if (authenticated == false)
            {
                InvalidateIdentity();
                return;
            }

            if (authenticated is true) IsAuthenticated = true;
            AccountEmail = string.IsNullOrEmpty(email) ? null : email;

            if (result is null)
            {
                HandleFailure();
                return;
            }

            if (!IsAuthenticated) IsAuthenticated = true;
            Current = result;
            lock (_lock) _lastSuccess = DateTime.UtcNow;
            SaveCache(result, _lastSuccess, key);
        }
        catch
        {
            HandleFailure();
        }
        finally
        {
            Interlocked.Exchange(ref _fetching, 0);
        }
    }

    private void HandleFailure()
    {
        lock (_lock)
        {
            if (DateTime.UtcNow - _lastSuccess > MaxStaleness)
                InvalidateIdentityUnlocked();
        }
    }

    private void InvalidateIdentity()
    {
        lock (_lock) InvalidateIdentityUnlocked();
    }

    private void InvalidateIdentityUnlocked()
    {
        Current = null;
        AccountEmail = null;
        IsAuthenticated = false;
        _accountKey = null;
        _lastSuccess = DateTime.MinValue;
        ClearCache();
    }

    // --- Disk cache

    /// <summary>AccountKey is Codex's local tokens.account_id from auth.json.</summary>
    private sealed record CachePayload(DateTime FetchedAt, string AccountKey, CodexUsageInfo Info);

    private static string CachePath
    {
        get
        {
            var baseDir = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
            if (string.IsNullOrEmpty(baseDir)) baseDir = Path.GetTempPath();
            return Path.Combine(baseDir, "AgentCord", "codex-usage-cache.json");
        }
    }

    private static CachePayload? LoadCache()
    {
        try
        {
            return JsonSerializer.Deserialize<CachePayload>(File.ReadAllText(CachePath));
        }
        catch
        {
            return null;
        }
    }

    private static void SaveCache(CodexUsageInfo info, DateTime fetchedAt, string accountKey)
    {
        try
        {
            var path = CachePath;
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            var tmp = path + ".tmp";
            File.WriteAllText(tmp, JsonSerializer.Serialize(new CachePayload(fetchedAt, accountKey, info)));
            File.Move(tmp, path, overwrite: true);
        }
        catch
        {
            // Best-effort cache.
        }
    }

    private static void ClearCache()
    {
        try { File.Delete(CachePath); }
        catch { }
        try { File.Delete(CachePath + ".tmp"); }
        catch { }
    }

    /// <summary>Stable ChatGPT account id from Codex's auth.json. Identity only —
    /// we never read or refresh the OAuth tokens ourselves.</summary>
    private static string? CurrentAccountKey()
    {
        try
        {
            var path = Path.Combine(CodexPaths.ResolveHome(), "auth.json");
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            if (!doc.RootElement.TryGetProperty("tokens", out var tokens)
                || tokens.ValueKind != JsonValueKind.Object)
                return null;
            var id = tokens.TryGetProperty("account_id", out var value) ? value.GetString() : null;
            return string.IsNullOrEmpty(id) ? null : id;
        }
        catch
        {
            return null;
        }
    }

    private static bool TryParseAccount(string line, out string? email, out bool authenticated)
    {
        email = null;
        authenticated = false;
        try
        {
            using var doc = JsonDocument.Parse(line);
            var root = doc.RootElement;
            if (!root.TryGetProperty("id", out var id)
                || !id.TryGetInt32(out var requestId)
                || requestId != 1)
                return false;

            if (!root.TryGetProperty("result", out var result)
                || result.ValueKind != JsonValueKind.Object
                || !result.TryGetProperty("account", out var account)
                || account.ValueKind != JsonValueKind.Object)
                return true;

            email = StringProp(account, "email");
            var type = StringProp(account, "type");
            authenticated = type is "chatgpt" or "personalAccessToken";
            return true;
        }
        catch
        {
            return false;
        }
    }

    private static bool TryParseResponse(string line, out CodexUsageInfo? info)
    {
        info = null;
        try
        {
            using var doc = JsonDocument.Parse(line);
            var root = doc.RootElement;
            if (!root.TryGetProperty("id", out var id)
                || !id.TryGetInt32(out var requestId)
                || requestId != 2)
                return false;

            if (!root.TryGetProperty("result", out var result)
                || result.ValueKind != JsonValueKind.Object
                || !result.TryGetProperty("rateLimits", out var limits)
                || limits.ValueKind != JsonValueKind.Object
                || !limits.TryGetProperty("primary", out var primary)
                || primary.ValueKind != JsonValueKind.Object)
                return true;

            var primaryWindow = ParseWindow(primary, limits.GetPropertyOrNull("rateLimitReachedType") is not null);
            var secondary = limits.TryGetProperty("secondary", out var secondaryElement)
                && secondaryElement.ValueKind == JsonValueKind.Object
                ? ParseWindow(secondaryElement, false)
                : null;

            var additional = new List<NamedUsageWindow>();
            if (result.TryGetProperty("rateLimitsByLimitId", out var byId)
                && byId.ValueKind == JsonValueKind.Object)
            {
                foreach (var item in byId.EnumerateObject().OrderBy(item => item.Name))
                {
                    if (item.Name.Equals("codex", StringComparison.OrdinalIgnoreCase)
                        || item.Value.ValueKind != JsonValueKind.Object)
                        continue;

                    var snapshot = item.Value;
                    var rawName = StringProp(snapshot, "limitName")
                        ?? StringProp(snapshot, "limitId")
                        ?? item.Name;
                    var label = DisplayName(rawName);
                    var reached = snapshot.GetPropertyOrNull("rateLimitReachedType") is not null;

                    if (snapshot.TryGetProperty("primary", out var scopedPrimary)
                        && scopedPrimary.ValueKind == JsonValueKind.Object)
                    {
                        additional.Add(new NamedUsageWindow
                        {
                            Id = $"{item.Name}-primary",
                            Label = label,
                            Window = ParseWindow(scopedPrimary, reached),
                        });
                    }
                    if (snapshot.TryGetProperty("secondary", out var scopedSecondary)
                        && scopedSecondary.ValueKind == JsonValueKind.Object)
                    {
                        additional.Add(new NamedUsageWindow
                        {
                            Id = $"{item.Name}-secondary",
                            Label = $"{label} · {WindowLabel(scopedSecondary, "Secondary")}",
                            Window = ParseWindow(scopedSecondary, false),
                        });
                    }
                }
            }

            info = new CodexUsageInfo
            {
                Primary = primaryWindow,
                PrimaryLabel = WindowLabel(primary, "Primary limit"),
                Secondary = secondary,
                SecondaryLabel = secondary is null ? null : WindowLabel(secondaryElement, "Secondary limit"),
                PlanType = StringProp(limits, "planType"),
                AdditionalWindows = additional,
            };
            return true;
        }
        catch
        {
            return false;
        }
    }

    private static UsageWindow ParseWindow(JsonElement window, bool reached)
    {
        var percent = NumberProp(window, "usedPercent") ?? 0;
        var rounded = Math.Clamp((int)Math.Round(percent), 0, 100);
        var resetSeconds = NumberProp(window, "resetsAt");
        return new UsageWindow
        {
            Percent = rounded,
            Severity = reached || rounded >= 90 ? "critical" : rounded >= 70 ? "warning" : "normal",
            ResetsAtMs = resetSeconds is null
                ? null
                : (long)(resetSeconds > 1_000_000_000_000 ? resetSeconds : resetSeconds * 1000),
        };
    }

    private static string WindowLabel(JsonElement window, string fallback)
    {
        var minutes = NumberProp(window, "windowDurationMins");
        if (minutes is null || minutes <= 0) return fallback;
        if (minutes <= 6 * 60) return "5-hour session";
        if (minutes <= 8 * 24 * 60) return "Weekly limit";
        if (minutes <= 40 * 24 * 60) return "Monthly limit";
        return fallback;
    }

    private static string DisplayName(string value) => string.Join(
        " ",
        value.Replace('_', ' ').Replace('-', ' ')
            .Split(' ', StringSplitOptions.RemoveEmptyEntries)
            .Select(word => char.ToUpperInvariant(word[0]) + word[1..]));

    private static string? StringProp(JsonElement obj, string name) =>
        obj.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static double? NumberProp(JsonElement obj, string name) =>
        obj.TryGetProperty(name, out var value)
        && value.ValueKind == JsonValueKind.Number
        && value.TryGetDouble(out var number)
            ? number
            : null;

    private static string? FindExecutable()
    {
        var candidates = new List<string>();
        if (Environment.GetEnvironmentVariable("CODEX_BINARY") is { Length: > 0 } configured)
            candidates.Add(configured);

        foreach (var dir in (Environment.GetEnvironmentVariable("PATH") ?? "")
                     .Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries))
        {
            candidates.Add(Path.Combine(dir.Trim('"'), "codex.exe"));
        }

        var local = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
        var profile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        candidates.Add(Path.Combine(local, "Programs", "OpenAI", "Codex", "bin", "codex.exe"));
        candidates.Add(Path.Combine(profile, ".local", "bin", "codex.exe"));

        return candidates.FirstOrDefault(File.Exists);
    }
}

internal static class JsonElementExtensions
{
    public static JsonElement? GetPropertyOrNull(this JsonElement element, string name) =>
        element.TryGetProperty(name, out var value)
        && value.ValueKind is not JsonValueKind.Null and not JsonValueKind.Undefined
            ? value
            : null;
}
